projects/drivers/util/fifo.c

```diff
@@ -0,0 +1,140 @@
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <string.h>
+#include <stdlib.h>
+#include "fifo.h"
+#include "error.h"
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief Create new fifo element
+ * @param buff - Data to be saved in fifo.
+ * @param len - Length of the data.
+ * @return fifo element in case of success, NULL otherwise
+ */
+static struct fifo * fifo_new_element(char *buff, uint32_t len)
+{
+	struct fifo *q = calloc(1, sizeof(struct fifo));
+	if (!q)
+		return NULL;
+
+	q->len = len;
+	q->data = calloc(1, len);
+	if (!(q->data)) {
+		free(q);
+		return NULL;
+	}
+	memcpy(q->data, buff, len);
+
+	return q;
+}
+
+/**
+ * @brief Get last element in fifo
+ * @param p_fifo - pointer to fifo
+ * @return fifo last element if exists, NULL otherwise
+ */
+static struct fifo *fifo_get_last(struct fifo *p_fifo)
+{
+	if(p_fifo == NULL)
+		return NULL;
+	while (p_fifo->next) {
+		p_fifo = p_fifo->next;
+	}
+
+	return p_fifo;
+}
+
+/**
+ * @brief Insert element to fifo, in the last position.
+ * @param pfifo - Pointer to fifo.
+ * @param buff - Data to be saved in fifo.
+ * @param len - Length of the data.
+ * @return SUCCESS in case of success, FAILURE otherwise
+ */
+int32_t fifo_insert_tail(struct fifo **p_fifo, char *buff, int32_t len)
+{
+	struct fifo *p, *q;
+
+	if (len <= 0)
+		return FAILURE;
+
+	q = fifo_new_element(buff, len);
+	if (!q)
+		return FAILURE;
+
+	if (!(*p_fifo)) {
+		*p_fifo = q;
+	} else {
+		p = fifo_get_last(*p_fifo);
+		p->next = q;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Remove fifo head
+ * @param pfifo - Pointer to fifo.
+ * @return next element in fifo if exists, NULL otherwise.
+ */
+struct fifo * fifo_remove_head(struct fifo *p_fifo)
+{
+	struct fifo *p = p_fifo;
+
+	if (p_fifo != NULL) {
+		p_fifo = p_fifo->next;
+		free(p->data);
+		p->len = 0;
+		p->next = NULL;
+		p->data = NULL;
```

Comment:
Are these still necessary if we `free(p)` later anyway?

Reply:
these don't count anymore.